src/vector/l_bfgs.rs

```diff
@@ -0,0 +1,233 @@
+///
+/// # fn main() {
+/// // Use the builder pattern to create a minimizer object
+/// // with default values for the parameters
+/// let min = LBFGSBuilder::default().build().unwrap();
+///
+/// // The function that is to be minimized, and an exact gradient function
+/// let center = arr1(&[0.9, 1.3, 0.5]);
+/// let f = |x: ArrayView1<f64>| (&x - &center).mapv(|xi| xi.powi(4)).scalar_sum();
+/// let x0 = Array1::ones(center.len());
+///
+/// // do the actual minimization
+/// let xmin = min.minimize_approx_grad(&f, x0.view());
+///
+/// println!("{:?}", xmin);
+/// assert!(xmin.all_close(&center, 1e-4))
+/// # }
+/// ```
+#[derive(Builder, Debug)]
+pub struct LBFGS {
+    /// The tolerance on the gradient used to terminate the optimization.
+    /// Smaller is more precise.
+    #[builder(default = "1e-12")]
+    pub gtol: f64,
+
+    /// The maximum number of iterations. The gradient is evaluated once per iteration.
+    /// Larger is more precise.
+    #[builder(default = "1500")]
+    pub max_iter: usize,
+
+    /// The number of datapoints to use to estimate the inverse hessian.
+    /// Larger is more precise. If `m` is larger than `x0.len()`, then
+    /// `x0.len()` is used.
+    #[builder(default = "5")]
+    pub m: usize,
+
+    /// The maximum step to be taken in the direction determined by the Quasi-Newton
+    /// method.
+    #[builder(default = "2.0")]
+    pub max_step: f64,
+
+    /// The tolerance on x used to terminate the line search.
+    /// Smaller is more precise.
+    #[builder(default = "1e-8")]
+    pub xtol: f64,
+
+    /// The maximum number of function evaluations.
+    #[builder(default = "1500")]
+    pub max_feval: usize,
+}
+
+impl LBFGS {
+    /// Minimize `func` starting in `x0` using a finite difference approximation
+    /// for the gradient. For more control over the approximation of the gradient,
+    /// you can use `minimize` with you own approximation.
+    pub fn minimize_approx_grad<F>(&self, func: F, x0: ArrayView1<f64>) -> Array1<f64>
+    where
+        F: Fn(ArrayView1<f64>) -> f64,
+    {
+        let eps = Array1::ones(x0.len()) * 1e-9;
+        let eps_view = eps.view();
+        let grad = |x: ArrayView1<f64>| ::utils::approx_fprime(x, &func, eps_view);
+        self.minimize(&func, &grad, x0)
+    }
+
+    /// Minimize `func` starting in `x0` using `grad` as the gradient of `func`.
+    pub fn minimize<F, G>(&self, func: F, grad: G, x0: ArrayView1<f64>) -> Array1<f64>
+    where
+        F: Fn(ArrayView1<f64>) -> f64,
+        G: Fn(ArrayView1<f64>) -> Array1<f64>,
+    {
+        let mut iter = 0;
+        let mut feval_count = 0;
+        let m = x0.len().min(self.m).max(1);
+
+        let mut hist = VecDeque::with_capacity(m);
+
+        let mut x = x0.to_owned();
+        let mut g = grad(x.view());
+
+        loop {
+            let dir = self.quasi_update(&g, &hist);
+            let a = {
+                let min = ::scalar::GoldenRatioBuilder::default()
```

Comment:
How about moving this minimizer implementation outside the loop? Not sure this works, but I don't see any obvious reasons why not.